src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -1,34 +1,101 @@
-                    .getPathInfo().substring(1))));
+            request.setAttribute("product", productDao.getProduct(parseProductId(request)));
+            request.setAttribute("cart", cartService.getCart(request).toString());
         } catch (NoSuchElementException e) {
             request.setAttribute("errorMessage", e.getMessage());
             request.getRequestDispatcher("/WEB-INF/pages/errorNoSuchElementException.jsp").forward(request, response);
         }
         request.getRequestDispatcher("/WEB-INF/pages/productDetails.jsp").forward(request, response);
+        session.setAttribute("message", "");
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        String stringQuantity = request.getParameter("quantity");
+        Long productId = parseProductId(request);
+        int quantity;
+        try {
+            NumberFormat formatter = NumberFormat.getInstance(request.getLocale());
+            quantity = Integer.parseInt(stringQuantity);
+        } catch (NumberFormatException e) {
+            request.setAttribute("errorMessage", "This field is for numbers only.");
+            doGet(request, response);
+            return;
+        }
+        if (quantity < 1) {
+            request.setAttribute("errorMessage", "Quantity must be a positive number.");
+            doGet(request, response);
+            return;
+        }
+        Cart cart = cartService.getCart(request);
+        try {
+            cartService.add(cart, productId, quantity);
+        } catch (OutOfStockException e) {
+            request.setAttribute("errorMessage", "Sorry, we don't have enough product stock! " +
+                    "Asked quantity: " + e.getRequestedQuantity() + ", available quantity: " + e.getAvailableQuantity());
+            doGet(request, response);
+            return;
+        }
+        HttpSession session = request.getSession();
+        session.setAttribute("message", "Product added successfully!");
+        response.sendRedirect(request.getContextPath() + "/products/" + productId);
+
+    }
+
+    private long parseProductId(HttpServletRequest request) throws NumberFormatException {
+        return Long.parseLong(request.getPathInfo().substring(1));
+    }
+
+    private void UpdateSearchHistory(HttpSession session, HttpServletRequest request) {
+        List<Product> recentProducts = (List<Product>) session.getAttribute("recentProducts");
+        if (recentProducts == null) {
+            recentProducts = new LinkedList<>();
+        }
+        recentProducts.removeIf(p -> p.getId() == parseProductId(request));
+
+        recentProducts.add(0, productDao.getProduct(parseProductId(request)));
+
+        if (recentProducts.size() > 3) {
+            recentProducts.remove(3);
+        }
+        session.setAttribute("recentProducts", recentProducts);
```

Comment:
Please move logic for processing recently viewed product in separate class + add tests